src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -278,8 +279,12 @@ OMapInnerNode::list(
 	      }
             }
             result.merge(std::move(child_result));
+	    if (result.size() == config.max_result_size) {
+	      return list_iertr::make_ready_future<seastar::stop_iteration>(
+		seastar::stop_iteration::yes);
+	    }
             ++iter;
-            assert(child_complete || result.size() == config.max_result_size);
```

Comment:
> Sorry for this inconvenience, the current implemetation of OMapManager::omap_list() may leak omap entries in the following scenario:
> 1. omap_list reached the last entry in some internal node;
> 2. omap_list hit max_result_size before exhausting all omap entries in that subtree;
>
> In this scenario, omap_list would return complete, and all of the entries beyond the end of the current list are leaked.

NP, I'm still not following what "leak" means, does it mean all of the entries beyond the end of the current list will be included even if it should be already complete?

Looking at how the logic are changed.
* The `child_complete == true` case here doesn't seem to be modified before and after this change (i.e. case 1).
* The `result.size() == config.max_result_size` case here (should be case 2):
  * before the change: repeat() will stop at the begin of the next iteration (L238 above) with complete == false.
  * after the change: repeat() will stop at L283 with the new logic, also with complete == false.

So, IIUC the only difference here is that the new logic skips a `++iter` in case of 2, and the results won't be changed.

 Probably I'm missing something from the above deduction?

Reply:
> > Sorry for this inconvenience, the current implemetation of OMapManager::omap_list() may leak omap entries in the following scenario:
> > 
> > 1. omap_list reached the last entry in some internal node;
> > 2. omap_list hit max_result_size before exhausting all omap entries in that subtree;
> > 
> > In this scenario, omap_list would return complete, and all of the entries beyond the end of the current list are leaked.
> 
> NP, I'm still not following what "leak" means, does it mean all of the entries beyond the end of the current list will be included even if it should be already complete?
> 
> Looking at how the logic are changed.
> 
> * The `child_complete == true` case here doesn't seem to be modified before and after this change (i.e. case 1).
> * The `result.size() == config.max_result_size` case here (should be case 2):
>   
>   * before the change: repeat() will stop at the begin of the next iteration (L238 above) with complete == false.
>   * after the change: repeat() will stop at L283 with the new logic, also with complete == false.
> 
> So, IIUC the only difference here is that the new logic skips a `++iter` in case of 2, and the results won't be changed.
> 
> Probably I'm missing something from the above deduction?

By "leak", I meant `omap_list` returns `complete` when it hit `max_result_size` and hasn't actually reached the end of the listing(`complete` means that all elements within [start, end) has been visited).

Reply:
> So, IIUC the only difference here is that the new logic skips a `++iter` in case of 2, and the results won't be changed.

No, the result's `complete` is changed.